lib/Runtime/Base/ThreadContext.h

```diff
@@ -107,16 +107,19 @@ class InterruptPoller
                 }\
         }
 
-#define LEAVE_SCRIPT_IF(scriptContext, condition, block) \
-        if (condition) \
+#define LEAVE_SCRIPT_IF_ACTIVE(scriptContext, externalCall) \
+        if (scriptContext->GetThreadContext()->IsScriptActive()) \
         { \
             BEGIN_LEAVE_SCRIPT(scriptContext); \
-            block \
+            externalCall \
             END_LEAVE_SCRIPT(scriptContext); \
         } \
         else \
         { \
-            block \
+            DECLARE_EXCEPTION_CHECK_DATA \
+            SAVE_EXCEPTION_CHECK \
+            externalCall \
+            RESTORE_EXCEPTION_CHECK \
         }
 
 #define ENTER_SCRIPT_IF(scriptContext, doCleanup, isCallRoot, hasCaller, condition, block) \
```

Comment:
I wonder why these  ENTER_SCRIPT_IF/LEAVE_SCRIPT_IF are introduced.   They seem limited to be used by Module.  

Is there are way to make this to always be in script or not with these calls by entering script earlier?

Reply:
The change came in as part of a combined PR: https://microsoft.visualstudio.com/OS/OS%20Team/_git/os/pullrequest/452680 that doesn't provide any rationale for it. 

I was going to extend the use of the macros to DynamicSourceHolder to fix OS bugs #17672089 and #17643917. My understanding in the case of DynamicSourceHolder that it's legitimate for the code to be executed outside of a script.